File: tyler/a1/credit.c
// include header
#include <cs50.h>
#include <stdio.h>

// functions
string validate_cc(long ccnum);
int sum_digits(int x);
string get_card_type(long ccnum);

//main
int main(void)
{
    // get cc number
    long ccnum = get_long("Number:");
    // validate the cc number and print the return
    printf("%s\n", validate_cc(ccnum));
}

string validate_cc(long ccnum)
{
    // cehck first if the number has the rught number of digits and leading digits to be valid
    string card_type = get_card_type(ccnum);
    
    if (card_type[0] == 'I')
    {
        // if the cc number is invalid at this point end here
        return card_type;
    }
    else
    {
        int i = 0; //ith digit off the cc number from right
        int sum_mult2_digits = 0; // sum of the products digits
        int sum_no_mult = 0; // sum of the other digits
        int digit; // placeholder for the current digit
        while (ccnum) // loop through all digits in the cc number
        {
            digit = ccnum % 10; // get the right most digit
            ccnum /= 10; // remove the right most digit
            if (i % 2 == 0) // check which sum to do
            {
                sum_no_mult += digit; // even digits from right (start a 0) get added
            }
            else
            {
                sum_mult2_digits += sum_digits(digit * 2); // ods get mult by 2 then have thirt digits summed 
            }
            i++;
            //printf("%ld, %d, %d, %d, %d \n", ccnum, digit, sum_no_mult, sum_mult2_digits, i);
        }
        if ((sum_no_mult + sum_mult2_digits) % 10 == 0) // check if the cc number if valid
        {
            return card_type; // if yes return card type
        }
        else
        {
            return "INVALID"; // if not return invalid
        }
    }
}

// sum the digits in the supplied integer
int sum_digits(int x)
{
    int sum = 0;
    while (x)
    {
        sum += x % 10; // add current digit
        x /= 10; // get next digit
    }
    return sum;
}

// determine what type of card this cc number corresponds to
string get_card_type(long ccnum)
{
    int first_digits = ccnum / 1000000000000;
    // visa, 13 or 16 digits starting with 4
    if (first_digits == 4 || first_digits / 1000 == 4)
    {
        return "VISA";
    }
    // amex, 15 digits starting with 34 or 37
    else if (first_digits / 10 == 34 || first_digits / 10 == 37)
    {
        return "AMEX";
    }
    // mastercard, 16 digits stating with 51 to 55
    else if (first_digits / 100 < 56 && first_digits / 100 >= 51)
    {
        return "MASTERCARD";        
    }
    else
    {
        return "INVALID";    
    }
}

File: tyler/a6/credit/credit.py
from cs50 import get_int
from math import log10

def luhn(n):
    odd = [ y if y < 10 else (y % 10) +1 for y in [(n // 10**x % 10)*2 for x in range(int(log10(n))+1) if x%2==1]]
    even = [(n // 10**x % 10) for x in range(int(log10(n))+1) if x%2==0]
    # print(odd, even)
    # print(sum(odd+even))
    return (sum(odd+even) % 10) == 0

n = get_int("Number: \n")
if luhn(n):
    if 4*10**12 <= n < 5*10**12 or 4*10**15 <= n < 5*10**15:
        print("VISA")
    elif 34*10**13 <= n < 35*10**13 or 37*10**13 <= n < 38*10**13:
        print("AMEX")
    elif 51*10**14 <= n < 56*10**14:
        print("MASTERCARD")
    else:
        print("INVALID")
else:
    print("INVALID")
        